### PR TITLE
add rt1308 and rt711 support

### PR DESCRIPTION
**Pull request by bardliao**
Add support for rt1308 and rt711. Tested on ICL with Realtek ALC711 AIOC TEST CARD.
We may want to use a better way to get the dai index, and "ASoC: SOF: intel: get dai index from dai name" can be a temporary solution.

### CODE REVIEW

File: sound/soc/intel/boards/cnl_rt700.c [hunk 1]
```diff
@@ -105,7 +105,8 @@ static const struct snd_soc_dapm_widget cnl_rt700_widgets[] = {
 static const struct snd_soc_dapm_route cnl_rt700_map[] = {
 	/*Headphones*/
 	{ "Headphones", NULL, "HP" },
-	{ "Speaker", NULL, "SPK" },
+	{ "Speaker", NULL, "SPOL" },
+	{ "Speaker", NULL, "SPOR" },
```
Comment (plbossart) [inline sound/soc/intel/boards/cnl_rt700.c]: I think this should be a different route (in a different structure), e.g. we need to keep the ability to test ALC700 only for on-board solutions, without RT1308. So I'd rather have a quirk and different routes really.

File: sound/soc/intel/boards/cnl_rt700.c [hunk 2]
```diff
@@ -148,6 +149,11 @@ SND_SOC_DAILINK_DEF(sdw0_pin,
 SND_SOC_DAILINK_DEF(sdw0_codec,
 	DAILINK_COMP_ARRAY(COMP_CODEC("sdw:0:25d:700:0:0", "rt700-aif1")));
 
+SND_SOC_DAILINK_DEF(sdw2_pin,
+	DAILINK_COMP_ARRAY(COMP_CPU("SDW2 Pin0")));
+SND_SOC_DAILINK_DEF(sdw2_codec,
+	DAILINK_COMP_ARRAY(COMP_CODEC("sdw:2:25d:1308:0:0", "rt1308-aif")));
+
```
Comment (plbossart) [inline sound/soc/intel/boards/cnl_rt700.c:209]: lets add sdw3 as well?

File: sound/soc/intel/boards/cnl_rt700.c [hunk 3]
```diff
@@ -178,6 +184,7 @@ SND_SOC_DAILINK_DEF(platform,
 		DAILINK_COMP_ARRAY(COMP_PLATFORM("0000:00:1f.3")));
 
 struct snd_soc_dai_link cnl_rt700_msic_dailink[] = {
+#if 1
```
Comment (plbossart) [inline sound/soc/intel/boards/cnl_rt700.c]: we should have different dailinks depending on whether we use the AOB or not.

File: sound/soc/intel/boards/cnl_rt700.c [hunk 4]
```diff
@@ -250,6 +248,13 @@ struct snd_soc_dai_link cnl_rt700_msic_dailink[] = {
 #endif
 };
 
+static struct snd_soc_codec_conf rt1308_codec_conf[] = {
+	{
+		.dev_name = "sdw:2:25d:1308:0:0",
+		.name_prefix = "rt1308",
```
Comment (plbossart) [inline sound/soc/intel/boards/cnl_rt700.c]: the prefix is not going to scale if you have 2 rt1308, so need to have rt1308-1 or something

File: sound/soc/codecs/rt711-sdw.c [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+		return -ENOMEM;
+
+	/* Initialize completion */
+	for (i = 0; i < num_of_ports; i++)
+		init_completion(&slave->port_ready[i]);
+
+	/* set the timeout values */
+	prop->clk_stop_timeout = 20;
+
+	return 0;
```
Comment (plbossart) [inline sound/soc/codecs/rt711-sdw.c:232]: this looks the same as the rt700 code? Is there any way we can factor common things between RT700 and RT711?

File: sound/soc/codecs/rt711-sdw.c [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+	return sdw_register_driver(&rt711_sdw_driver);
+}
+
+static void __exit sdw_slave_exit(void)
+{
+	sdw_unregister_driver(&rt711_sdw_driver);
+}
+
+module_init(sdw_slave_init);
+module_exit(sdw_slave_exit);
```
Comment (plbossart) [inline sound/soc/codecs/rt711-sdw.c:330]: same comments, use sdw_module_driver.

File: sound/soc/intel/boards/cnl_rt700.c [hunk 7]
```diff
@@ -149,7 +149,7 @@ static int cnl_rt700_codec_fixup(struct snd_soc_pcm_runtime *rtd,
 SND_SOC_DAILINK_DEF(sdw0_pin,
 	DAILINK_COMP_ARRAY(COMP_CPU("SDW0 Pin0")));
 SND_SOC_DAILINK_DEF(sdw0_codec,
-	DAILINK_COMP_ARRAY(COMP_CODEC("sdw:0:25d:700:0:0", "rt700-aif1")));
+	DAILINK_COMP_ARRAY(COMP_CODEC("sdw:0:25d:711:0:1", "rt711-aif1")));
```
Comment (plbossart) [inline sound/soc/intel/boards/cnl_rt700.c]: this should be added and used inside a different dailink when we use the AOC

File: sound/soc/codecs/rt1308-sdw.c [hunk 8]
```diff
@@ -0,0 +1,779 @@
...
+					 GFP_KERNEL);
+	if (!slave->port_ready)
+		return -ENOMEM;
+
+	/* Initialize completion */
+	for (i = 0; i < num_of_ports; i++)
+		init_completion(&slave->port_ready[i]);
+
+	/* set the timeout values */
+	prop->clk_stop_timeout = 20;
```
Comment (plbossart) [inline sound/soc/codecs/rt1308-sdw.c:330]: same copy/paste it seems. maybe we can have a common function between all rtk codecs for inits.

**Review by plbossart — CHANGES_REQUESTED** [review]
Thanks @bardliao I think the main issue is that we want to keep functionality on the default ICL boards, as well as support the AOC board, so I think we need to have different quirks and dailinks rather than a configuration that only supports AOC.
For the dai index and grouping will look at this separately. Thanks for providing the pointers.

File: sound/soc/intel/boards/cnl_rt700.c (hunk 2 above)
Comment (bardliao) [reply sound/soc/intel/boards/cnl_rt700.c:209]: @plbossart We don't have rt715 codec driver for now. So, the sound card will not probed if we add sdw3.

**Comment by bardliao** [comment]
> Thanks @bardliao I think the main issue is that we want to keep functionality on the default ICL boards, as well as support the AOC board, so I think we need to have different quirks and dailinks rather than a configuration that only supports AOC.

@plbossart I squashed some patches and use module param as a quirk to define if AOC is used. I didn't touch codec driver part. Maybe @shumingfan can chime in?

**Pull request closed without merging**